qcdb/parsers/parse.py

```diff
@@ -50,10 +57,78 @@ def get_library_read_type(self, library_read_type):
         if library_read_type == '2':
             return 'paired-end reverse'
         if library_read_type.isalpha():
-            return 'single ended' 
+            return 'single ended'
         else:
             raise KeyError("Library read type not recognized.")
 
+    # get the reference map for this qc_program (e.g. {kmercount: {long_name: l}})
+    def get_reference_map(self):
+        map = {}
+        s = select([self.ref_table.c.qc_metric, self.ref_table.c.field_name, self.ref_table.c.field_code]).where(self.ref_table.c.qc_program == self.qc_program)
+        for row in self.session.execute(s):
+            # add metric to map if it doesn't yet exist
+            if row['qc_metric'] not in map:
+                map[row['qc_metric']] = {}
+
+            map[row['qc_metric']][row['field_name']] = row['field_code']
+
+        return map
+
+    # utility function to create field code
+    def get_new_mapping(self, field_name, codes):
+        print("Creating new code for ", field_name, " these codes already exist:", codes)
+        s = field_name.lower()
+        s = re.sub('[^0-9a-zA-Z]+', ' ', s)
+        xl = s.split()
+        s = re.sub('[^0-9a-zA-Z]+', '', s)
+        first_char = 'x'
+        for c in s:
+            if c.isalpha():
+                first_char = c
+                break
+
+        # return first alpha if not used
+        if first_char not in codes:
+            return first_char
+
+        candidate = first_char
+        i = 1
+        while i < len(xl):
+            candidate += xl[i][0]
+            i+=1
+
+        # return acronym if not used
+        if candidate not in codes:
+            return candidate
+
+        # add counter to acronym until not used
+        cnt = 1
+        while True:
+            if candidate + str(cnt) not in codes:
+                return candidate + str(cnt)
+            cnt += 1
+
+    # create new mapping value from field name to code
+    def get_mapped_val(self, metric_type, field_name):
+        if not self.build_ref:
+            raise Exception('Can only update reference mappings when buildref set to true')
+
+        if metric_type not in self.ref_map:
+            self.ref_map[metric_type] = {}
+
+        # section of the map we need to guaranee uniqueness over
+        mmap = self.ref_map[metric_type]
+
+        if field_name in mmap:
+            return mmap[field_name]
+        else:
+            new_val = self.get_new_mapping(field_name, mmap.values())
+            self.ref_map[metric_type][field_name] = new_val
+            self.session.execute(self.ref_table.insert(), [{'qc_program':self.qc_program, 'qc_metric':metric_type, 'field_name':field_name, 'field_code':new_val, 'display_name':field_name}])
```

Comment:
Since `display_name` is just the `field_name`, shouldn't we delete it?

Reply:
This would be a manual database update so that something other than the field_name could be displayed on the front end.  I default-populated it with the field_name so it wasn't null.  I think this was a feature request from Ashok.